### PR TITLE
chore: update Knip to v5

### PR DESCRIPTION
**Pull request by webpro**
When introducing Knip ~2 years ago I think we've kinda implicitly concluded Knip v1 wasn't ready for a repository like this. Too much configuration and knowledge was required for to use it in a convenient and productive way.

We've come a long way. I think now that we have Knip v5 it's useful even (or especially?!) for a large repo like freeCodeCamp. I would even say it could be useful to add to CI. But for now I've kept the current strategy of using `npx -y knip` intact to not introduce any surprises.

The biggest/most relevant changes in Knip:

- All workspaces are analyzed by default (in ~23s on my machine)
- [Support for CommonJS](https://knip.dev/guides/working-with-commonjs) has improved a lot, but still a few modification where necessary to remove ambiguity (also prepares for a potential migration to ESM).

This repo has been super useful to improve Knip itself, so I'm grateful for that and the least I could do in return is to actually remove some clutter from the freeCodeCamp repo.

I'm going to add freeCodeCamp to the integration test suite of Knip itself to prevent future regressions (less surprises for the maintainers in case Knip would be used more often).

~Before un-drafting this PR I'll make sure QA is green.~ 

**Edit 09/24**: Looking at the resulting config ([knip.jsonc](https://github.com/freeCodeCamp/freeCodeCamp/blob/1cb5c864e393bf20096cca7a41c9bd0f68d5520e/knip.jsonc)), nearly all of it is to compensate for custom structures/issues that can't really be automatically accounted for by a generic project linter. That is to say, I think this is basically as good as it gets (which could still not be worth managing this configuration, that's up to you obviously, no hard feelings either way).

---

Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #XXXXX

### CODE REVIEW

File: client/src/assets/icons/b2-english.tsx
Comment (ilenia-magoni) [inline client/src/assets/icons/b2-english.tsx:1]: we need this, even if it's not yet used. There are assets for WIP parts of the curriculum

File: client/src/assets/icons/c2-english.tsx
Comment (ilenia-magoni) [inline client/src/assets/icons/c2-english.tsx:1]: and also we need this

File: client/src/assets/icons/dot-net-logo.tsx
Comment (ilenia-magoni) [inline client/src/assets/icons/dot-net-logo.tsx:1]: I don't know where this is planned to use, but if it's there I would guess we also need this

**Comment by ilenia-magoni** [comment]
please do not delete the icons

**Comment by webpro** [comment]
> please do not delete the icons

Sure, the files have been reverted + ignored.

**Review by raisedadead — APPROVED** [review]
Thanks for the PR @webpro - We definitely like `Knip`. Happy to see how far it has come. Good luck!

**Review by ojeytonwilliams — APPROVED** [review]
Thanks for taking care of the migration, @webpro. Hopefully one day the repo won't be such a pain to configure, but at least we're a useful test case :cold_sweat:

**Comment by webpro** [comment]
You're more than welcome, thanks for merging.

Now that it's in a clean state, you could run `pnpm run knip:all` every once in a while to keep tabs on the state of unused files, exports and dependencies. Or even add it to CI once you're confident it's stable enough.

Best of luck!

**Comment by webpro** [comment]
(FYI, it's added to knip's integration suite. Not much to see here but it's a great repo to keep exercising: https://github.com/webpro-nl/knip/actions/runs/11054492711/job/30711431413)

**Comment by ojeytonwilliams** [comment]
Oh, cool. So, the idea is that you want to know if knip starts throwing unexpectedly?

**Comment by webpro** [comment]
Indeed. But also I'd like to know if it starts to report false positives. I wouldn't know without actually go in and review the Action job's output, but the goal is to remove `--no-exit-code`.

**Comment by ojeytonwilliams** [comment]
Tricky. Maybe the solution is just for us to try using it in our CI. Then we can just @ you if we don't understand what's going on. How does that sound?

I'd like to give it another go, so I'll add it now and we can see how it works in practice.

**Comment by webpro** [comment]
> Tricky. Maybe the solution is just for us to try using it in our CI. Then we can just @ you if we don't understand what's going on. How does that sound?

Absolutely, just don't want to cause any extra work on your end. But if you decide to do so, both projects can benefit.

> I'd like to give it another go, so I'll add it now and we can see how it works in practice.
> 
> EDIT: I spoke too soon, it looks like there's still a lot to remove :(

What's the command you're using? A `pnpm run knip:all` in main currently gives no issues reported by Knip here. But maybe you have a different repo state than mine with regards to what exactly's being built, etc? Here are the steps I've used: https://github.com/webpro-nl/knip/blob/2b550b0f8e384f903cc8bb7d0e7541d04df0547a/.github/workflows/integration.yml#L126-L130

**Comment by ojeytonwilliams** [comment]
After following those steps I got this:

<details>
<summary> Knip output </summary>

Unused files (53)
jest.transform.js
tools/challenge-auditor/index.ts
tools/scripts/sync-i18n.ts
tools/scripts/lint/validate-keys.ts
api/src/server.ts
client/i18n/schema-validation.ts
client/tools/create-env.ts
client/tools/download-trending.ts
client/tools/read-env.ts
client/tools/schema/trending-schema.ts
curriculum/lint-localized.js
tools/challenge-helper-scripts/create-empty-steps.ts
tools/challenge-helper-scripts/create-next-challenge.ts
tools/challenge-helper-scripts/create-next-step.ts
tools/challenge-helper-scripts/create-next-task.ts
tools/challenge-helper-scripts/create-project.ts
tools/challenge-helper-scripts/create-this-challenge.ts
tools/challenge-helper-scripts/delete-challenge.ts
tools/challenge-helper-scripts/delete-step.ts
tools/challenge-helper-scripts/delete-task.ts
tools/challenge-helper-scripts/insert-challenge.ts
tools/challenge-helper-scripts/insert-step.ts
tools/challenge-helper-scripts/insert-task.ts
tools/challenge-helper-scripts/rename-challenge-files.ts
tools/challenge-helper-scripts/reorder-tasks.ts
tools/challenge-helper-scripts/repair-meta.ts
tools/challenge-helper-scripts/update-challenge-order.ts
tools/challenge-helper-scripts/update-step-titles.ts
tools/challenge-helper-scripts/helpers/get-challenge-template.ts
tools/challenge-helper-scripts/helpers/get-last-step-file-number.ts
tools/challenge-helper-scripts/helpers/new-challenge-prompts.ts
tools/challenge-helper-scripts/helpers/new-task-prompts.ts
tools/scripts/seed/seed-demo-user.js
tools/scripts/seed/seed-ms-username.js
tools/scripts/seed/seed-surveys.js
tools/scripts/seed/user-data.js
tools/scripts/seed-exams/create-exams.js
tools/scripts/seed-exams/exam-schema.js
tools/challenge-editor/api/server.ts
tools/challenge-editor/api/configs/paths.ts
tools/challenge-editor/api/configs/super-block-list.ts
tools/challenge-editor/api/interfaces/partial-meta.ts
tools/challenge-editor/api/interfaces/tools.ts
tools/challenge-editor/api/routes/block-route.ts
tools/challenge-editor/api/routes/index-route.ts
tools/challenge-editor/api/routes/save-route.ts
tools/challenge-editor/api/routes/step-route.ts
tools/challenge-editor/api/routes/super-block-route.ts
tools/challenge-editor/api/routes/tools-route.ts
tools/challenge-editor/api/utils/get-blocks.ts
tools/challenge-editor/api/utils/get-step-contents.ts
tools/challenge-editor/api/utils/get-steps.ts
tools/challenge-editor/api/utils/save-step.ts
Unused dependencies (3)
cors         tools/challenge-editor/api/package.json
express      tools/challenge-editor/api/package.json
gray-matter  tools/challenge-editor/api/package.json
Unused devDependencies (21)
@types/js-yaml     client/package.json                        
@types/node-fetch  client/package.json                        
joi                client/package.json                        
js-yaml            client/package.json                        
node-fetch         client/package.json                        
glob               curriculum/package.json                    
babel-jest         package.json                               
@types/cors        tools/challenge-editor/api/package.json    
@types/express     tools/challenge-editor/api/package.json    
dotenv             tools/challenge-editor/api/package.json    
@types/inquirer    tools/challenge-helper-scripts/package.json
inquirer           tools/challenge-helper-scripts/package.json
prettier           tools/challenge-helper-scripts/package.json
debug              tools/scripts/seed-exams/package.json      
dotenv             tools/scripts/seed-exams/package.json      
joi                tools/scripts/seed-exams/package.json      
joi-objectid       tools/scripts/seed-exams/package.json      
mongodb            tools/scripts/seed-exams/package.json      
debug              tools/scripts/seed/package.json            
dotenv             tools/scripts/seed/package.json            
mongodb            tools/scripts/seed/package.json
Unused exports (9)
PORT                     unknown  api/src/utils/env.ts:121:14                      
HOST                     unknown  api/src/utils/env.ts:122:14                      
insertStep               unknown  tools/challenge-helper-scripts/commands.ts:107:21
createEmptySteps         unknown  tools/challenge-helper-scripts/commands.ts:107:33
deleteStep               unknown  tools/challenge-helper-scripts/commands.ts:107:9 
updateTaskMeta           unknown  tools/challenge-helper-scripts/utils.ts:220:20   
updateTaskMarkdownFiles  unknown  tools/challenge-helper-scripts/utils.ts:221:18   
insertChallengeIntoMeta  unknown  tools/challenge-helper-scripts/utils.ts:223:21   
deleteChallengeFromMeta  unknown  tools/challenge-helper-scripts/utils.ts:225:22
Unused exported types (1)
Meta  type  tools/challenge-helper-scripts/helpers/project-metadata.ts:6:13

</details>

It's definitely possible my repo is an a weird state, but it's behaving itself otherwise.

**Comment by webpro** [comment]
Dang, that's odd. If you're willing to do a `git clean -fdx` we could compare:

<details>
Screenshot 2024-09-27 at 11 27 59 AM
</details>

Also, I now see that I've accidentally added https://github.com/freeCodeCamp/freeCodeCamp/blob/9c6ab8de9b36df8bb72e01ed466973d45b37d62b/package.json#L132

Sorry about that. Could this be the issue?

**Comment by ojeytonwilliams** [comment]
Doesn't seem to be, no. I removed it (and git cleaned), but nothing much changed. I'm a bit mystified, to be honest.

It works fine in Gitpod, so _something_ is wrong with my local setup. I'll let you know if/when I figure it out!

**Pull request merged**